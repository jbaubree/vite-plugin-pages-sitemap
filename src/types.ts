/**
 * Plugin options.
 */
interface Options {
  /**
   * Base URI
   * @default 'http://localhost/'
   */
  hostname: string
  /**
   * Generated routes from vite-plugin-pages
   * @default []
   */
  routes: any[]
  /**
   * File name
   * @default 'sitemap'
   */
  filename: string
  /**
   * Destination path
   * @default 'public'
   */
  dest: string
  /**
   * Change frequency option for sitemap
   * @default 'daily'
   */
  changefreq: string
  /**
   * Priority option for sitemap
   * @default 1
   */
  priority: number
  /**
   * Last modification option for sitemap
   * @default new Date()
   */
  lastmod: Date
  /**
   * Converts XML into a human readable format
   * @default false
   */
  readable: boolean
  /**
   * Is Nuxt.js style route naming
   * @default false
   */
  nuxtStyle: boolean
}

export type UserOptions = Partial<Options>

export interface ResolvedOptions extends Options {}
