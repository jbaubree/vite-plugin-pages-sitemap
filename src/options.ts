import { ResolvedOptions, UserOptions } from './types'

export function resolveOptions(userOptions: UserOptions): ResolvedOptions {
  return Object.assign(
    {
      hostname: 'http://localhost/',
      routes: [],
      filename: 'sitemap',
      dest: 'public',
      changefreq: 'daily',
      priority: 1,
      lastmod: new Date(),
      nuxtStyle: false,
    },
    userOptions,
  )
}
