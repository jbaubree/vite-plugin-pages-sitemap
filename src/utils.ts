const dynamicRouteRE = /\[.+\]/
const nuxtDynamicRouteRE = /_[\s\S]*/

export function isDynamicRoute(routePath: string, nuxtStyle: boolean) {
  if (!routePath)
    return false
  return nuxtStyle
    ? nuxtDynamicRouteRE.test(routePath)
    : dynamicRouteRE.test(routePath)
}

export function ensureSuffix(suffix: string, str: string) {
  if (!str.endsWith(suffix))
    return str + suffix
  return str
}

export function removeMaybeSuffix(suffix: string, str: string) {
  if (!str.endsWith(suffix))
    return str
  return str.slice(0, -suffix.length)
}
