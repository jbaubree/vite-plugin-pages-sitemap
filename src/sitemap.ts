import { writeFileSync } from 'fs'
import { resolve } from 'path'
import { ensurePrefix } from '@antfu/utils'
import format from 'xml-formatter'

import { ensureSuffix, isDynamicRoute, removeMaybeSuffix } from './utils'
import type { ResolvedOptions } from './types'

export function getSitemapLinks(options: ResolvedOptions) {
  const hostname = options.hostname
  const routes = flatRoutes([...options.routes.values()])
  return routes
    .filter(pageRoute => !isDynamicRoute(pageRoute.component as string, options.nuxtStyle) || typeof pageRoute === 'string')
    .filter(pageRoute => !options.exclude.includes(pageRoute.path))
    .map(pageRoute => ({
      url: `${removeMaybeSuffix('/', hostname)}${typeof pageRoute === 'string' ? pageRoute : pageRoute.path}`,
      changefreq: options.changefreq,
      priority: options.priority,
      lastmod: options.lastmod,
    }))
}

export function flatRoutes(routes: any[]) {
  const flatRoutes: Set<{ component?: string; path: string }> = new Set()

  const getFlatRoutes = (routes: any[], prefix = '') => {
    prefix = prefix.replace(/\/$/g, '')
    for (const route of routes) {
      if (typeof route === 'string') {
        flatRoutes.add({
          component: undefined,
          path: route,
        })
      }
      else {
        let path = route.path
        if (route.path) {
          path = prefix && !route.path.startsWith('/')
            ? `${prefix}/${route.path}`
            : route.path

          flatRoutes.add({
            component: route.component,
            path,
          })
        }
        if (Array.isArray(route.children))
          getFlatRoutes(route.children, path)
      }
    }
  }

  getFlatRoutes(routes)
  return [...flatRoutes]
}

export function getDestPath(dest: string) {
  return ensurePrefix('./', dest)
}

export function getResolvedPath(filename: string, extension: string, dest: string) {
  const filenameWithExtension = ensureSuffix(extension, filename)
  return resolve(`${getDestPath(dest)}/${filenameWithExtension}`)
}

export function writeXmlFile(resolvedPath: string, str: string, options: ResolvedOptions) {
  writeFileSync(resolvedPath, options.readable ? format(str) : str)
}

export function writeRobotFile(resolvedPath: string, options: ResolvedOptions) {
  const str = 'User-agent: *\n'
    .concat(`${options.allowRobots ? 'Allow' : 'Disallow'}: /\n\n`)
    .concat(`Sitemap: ${getFinalSitemapPath(options)}`)
  writeFileSync(resolvedPath, str)
}

export function getFinalSitemapPath(options: ResolvedOptions) {
  return `${ensureSuffix('/', options.hostname)}${ensureSuffix('.xml', options.filename)}`
}
