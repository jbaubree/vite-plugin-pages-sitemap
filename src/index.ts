import { createWriteStream, existsSync, mkdirSync } from 'fs'
import { resolve } from 'path'
import { SitemapStream } from 'sitemap'
import { ensurePrefix } from '@antfu/utils'

import { ensureSuffix, isDynamicRoute, removeMaybeSuffix } from './utils'
import { resolveOptions } from './options'
import type { ResolvedOptions, UserOptions } from './types'

export default function generateSitemap(options: UserOptions) {
  const resolvedOptions: ResolvedOptions = resolveOptions(options)
  if (!resolvedOptions.routes.length) return
  if (!existsSync(getDestPath(resolvedOptions))) mkdirSync(getDestPath(resolvedOptions))

  const stream = new SitemapStream()
  const writeStream = createWriteStream(getResolvedPath(resolvedOptions))
  getSitemapLinks(resolvedOptions).forEach(item => stream.write(item))
  stream.pipe(writeStream)
  stream.end()
}

export function getSitemapLinks(options: ResolvedOptions) {
  const hostname = options.hostname

  return [...options.routes.values()]
    .filter(pageRoute => !isDynamicRoute(pageRoute.component as string, options.nuxtStyle))
    .map(pageRoute => ({
      url: removeMaybeSuffix('/', hostname) + pageRoute.path,
      changefreq: options.changefreq,
      priority: options.priority,
      lastmod: options.lastmod,
    }))
}

export function getDestPath(sitemap: ResolvedOptions) {
  return ensurePrefix('./', sitemap.dest)
}

export function getResolvedPath(sitemap: ResolvedOptions) {
  const SITEMAP_EXTENSION = '.xml'
  const filenameWithExtension = ensureSuffix(SITEMAP_EXTENSION, sitemap.filename)
  return resolve(`${getDestPath(sitemap)}/${filenameWithExtension}`)
}
