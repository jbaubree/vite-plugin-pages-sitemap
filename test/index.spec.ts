import { resolve } from 'path'
import { existsSync, readFileSync } from 'fs'
import { describe, expect, test } from 'vitest'
import format from 'xml-formatter'

import generateSitemap, { getDestPath, getFinalSitemapPath, getResolvedPath, getSitemapLinks, writeRobotFile, writeXmlFile } from '../src'
import { resolveOptions } from '../src/options'

describe('Index', () => {
  test('Get sitemap links', async () => {
    expect(getSitemapLinks(resolveOptions({}))).toEqual([])
    expect(getSitemapLinks(resolveOptions({
      routes: ['/route'],
    }))).toMatchInlineSnapshot([{
      lastmod: expect.any(Date),
    }], `
      [
        {
          "changefreq": "daily",
          "lastmod": Any<Date>,
          "priority": 1,
          "url": "http://localhost/route",
        },
      ]
    `)
    expect(getSitemapLinks(resolveOptions({
      routes: ['/route'],
      exclude: ['/route'],
    }))).toMatchInlineSnapshot([])
    expect(getSitemapLinks(resolveOptions({
      routes: [{
        path: '/route',
        component: '/src/pages/route/index.vue',
      }],
    }))).toMatchInlineSnapshot([{
      lastmod: expect.any(Date),
    }], `
      [
        {
          "changefreq": "daily",
          "lastmod": Any<Date>,
          "priority": 1,
          "url": "http://localhost/route",
        },
      ]
    `)
    expect(getSitemapLinks(resolveOptions({
      routes: [{
        path: '/route',
        component: '/src/pages/route/index.vue',
        children: [{
          path: 'nested',
          component: '/src/pages/route/nested/index.vue',
        }],
      }],
    }))).toMatchInlineSnapshot([{
      lastmod: expect.any(Date),
    }, {
      lastmod: expect.any(Date),
    }], `
      [
        {
          "changefreq": "daily",
          "lastmod": Any<Date>,
          "priority": 1,
          "url": "http://localhost/route",
        },
        {
          "changefreq": "daily",
          "lastmod": Any<Date>,
          "priority": 1,
          "url": "http://localhost/route/nested",
        },
      ]
    `)
    expect(getSitemapLinks(resolveOptions({
      routes: [{
        path: '/route/:id',
        component: '/src/pages/route/[id].vue',
      }],
    }))).toEqual([])
  })

  test('Get dest path', async () => {
    expect(getDestPath('public')).toEqual('./public')
    expect(getDestPath('./public')).toEqual('./public')
  })

  test('Get resolved path', async () => {
    expect(getResolvedPath('sitemap', '.xml', 'public')).toEqual(resolve('./public/sitemap.xml'))
    expect(getResolvedPath('sitemap.xml', '.xml', 'public')).toEqual(resolve('./public/sitemap.xml'))
    expect(getResolvedPath('sitemap.xml', '.xml', './public')).toEqual(resolve('./public/sitemap.xml'))
  })

  test('Generate sitemap', async () => {
    generateSitemap({})
    expect(existsSync(getDestPath('public'))).toBe(false)
    const options = {
      dest: 'test/sitemap',
      routes: [
        {
          name: 'index',
          path: '/',
          component: '/src/pages/index.vue',
          props: true,
        },
      ],
      allowRobots: false,
    }
    generateSitemap(options)
    expect(existsSync(getDestPath('test/sitemap'))).toBe(true)
  })

  test('Write XML file', async () => {
    const TEMPLATE = '<?xml version="1.0" encoding="UTF-8"?><urlset></urlset>'
    writeXmlFile(resolve('./test/sitemap/sitemap.xml'), TEMPLATE, resolveOptions({}))
    expect(readFileSync(resolve('./test/sitemap/sitemap.xml')).toString('utf-8')).toEqual(TEMPLATE)
    writeXmlFile(resolve('./test/sitemap/sitemap.xml'), TEMPLATE, resolveOptions({ readable: true }))
    expect(readFileSync(resolve('./test/sitemap/sitemap.xml')).toString('utf-8')).toEqual(format(TEMPLATE))
  })

  test('Write robots file', async () => {
    writeRobotFile(resolve('./test/sitemap/robots.txt'), resolveOptions({}))
    expect(readFileSync(resolve('./test/sitemap/robots.txt')).toString('utf-8')).toEqual(
      'User-agent: *\nAllow: /\n\nSitemap: http://localhost/sitemap.xml',
    )
    writeRobotFile(resolve('./test/sitemap/robots.txt'), resolveOptions({ allowRobots: false }))
    expect(readFileSync(resolve('./test/sitemap/robots.txt')).toString('utf-8')).toEqual(
      'User-agent: *\nDisallow: /\n\nSitemap: http://localhost/sitemap.xml',
    )
  })

  test('Get final sitemap path', async () => {
    expect(getFinalSitemapPath(resolveOptions({}))).toEqual(
      'http://localhost/sitemap.xml',
    )
    expect(getFinalSitemapPath(resolveOptions({
      hostname: 'http://test.com/',
      filename: 'test.xml',
    }))).toEqual(
      'http://test.com/test.xml',
    )
  })
})
