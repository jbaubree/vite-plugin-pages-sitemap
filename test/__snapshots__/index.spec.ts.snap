// Vitest Snapshot v1

exports[`Index > Generate sitemap 1`] = `undefined`;
